nerfacc/cuda/csrc/grid.cu

```diff
@@ -305,80 +356,127 @@ std::vector<RaySegmentsSpec> traverse_grids(
     dim3 threads = dim3(min(max_threads, n_rays));
     dim3 blocks = dim3(min(max_blocks, ceil_div<int32_t>(n_rays, threads.x)));
 
-    // Sort the intersections. [n_rays, n_grids * 2]
-    torch::Tensor t_sorted, t_indices;
-    if (n_grids > 1) {
-        std::tie(t_sorted, t_indices) = torch::sort(torch::cat({t_mins, t_maxs}, -1), -1);
-    }
-    else {
-        t_sorted = torch::cat({t_mins, t_maxs}, -1);
-        t_indices = torch::arange(
-            0, n_grids * 2, t_mins.options().dtype(torch::kLong)
-        ).expand({n_rays, n_grids * 2}).contiguous();
-    }
-    
     // outputs
     RaySegmentsSpec intervals, samples;
+    torch::Tensor terminate_planes;
+    if (compute_terminate_planes) 
+        terminate_planes = torch::empty({n_rays}, rays_o.options());
+
+    if (over_allocate) {
+        // over allocate the memory so that we can traverse the grids in a single pass.
+        int total_steps = n_rays * traverse_steps_limit;
+        if (compute_intervals) {
+            intervals.chunk_cnts = torch::full({n_rays}, -1, rays_o.options().dtype(torch::kLong));
+            intervals.chunk_starts = torch::arange(
+                0, total_steps * 2, traverse_steps_limit * 2, rays_o.options().dtype(torch::kLong));
+            intervals.memalloc_data(total_steps * 2, true, true);
+        }
+        if (compute_samples) {
+            samples.chunk_cnts = torch::full({n_rays}, -1, rays_o.options().dtype(torch::kLong));
+            samples.chunk_starts = torch::arange(
+                0, total_steps, traverse_steps_limit, rays_o.options().dtype(torch::kLong));
+            samples.memalloc_data(total_steps, false, true, true);
```

Comment:
When in test mode, if we have the full `n_rays` each time (which will be contained within a while loop), memory allocation will become excessive due to the increase of `traverse_steps_limit` within the loop. However, if we only pass alive rays and not the full `n_rays`, then the `ray_mask` becomes redundant as it only indicates whether or not a ray is alive.

Reply:
Good points. I didn't think of that. 

So how about we do this -- We get rid of the `rays_mask`, and consider all rays that being passed to C are alive rays. And when we use the `traverse_grids` function in python, we pass it only the alive rays by masking out others. Something like:

```
travers_grids(rays_o[masks], rays_d[masks] ...)
```

Reply:
resolved in the latest commit